src/libraries/System.Linq/src/System/Linq/Max.cs

```diff
@@ -8,157 +8,70 @@ namespace System.Linq
 {
     public static partial class Enumerable
     {
-        public static int Max(this IEnumerable<int> source)
-        {
-            if (source.TryGetSpan(out ReadOnlySpan<int> span))
-            {
-                return Max(span);
-            }
+        public static int Max(this IEnumerable<int> source) => MaxInteger(source);
 
-            int value;
-            using (IEnumerator<int> e = source.GetEnumerator())
-            {
-                if (!e.MoveNext())
-                {
-                    ThrowHelper.ThrowNoElementsException();
-                }
+        public static int? Max(this IEnumerable<int?> source) => MaxInteger(source);
 
-                value = e.Current;
-                while (e.MoveNext())
-                {
-                    int x = e.Current;
-                    if (x > value)
-                    {
-                        value = x;
-                    }
-                }
-            }
+        public static long Max(this IEnumerable<long> source) => MaxInteger(source);
 
-            return value;
-        }
+        public static long? Max(this IEnumerable<long?> source) => MaxInteger(source);
 
-        private static int Max(ReadOnlySpan<int> span)
+        private static T MaxInteger<T>(this IEnumerable<T> source) where T : struct, IBinaryInteger<T>
```

Comment:
Do we want to log an issue proposing any of these generic forms as new API surface?

Reply:
There's https://github.com/dotnet/runtime/issues/64031.

But as written this isn't actually scalable to any T that meets the constraint, as it tries to use `Vector<T>` and that's limited to just some `T`s.  What's the recommended way to special-case the use of vectorization?  Does `Vector<T>` or `Vector128<T>` provide an IsSupported that factors in the `T` type?  I couldn't find one but may have missed it.

Reply:
We have `Vector.IsHardwareAccelerated` but nothing that currently says if a given `T` is supported. That wouldn't be difficult to expose, particularly if it helped scenarios like this.

Reply:
Seems like adding something here would be helpful.

Reply:
I'll log an issue for this